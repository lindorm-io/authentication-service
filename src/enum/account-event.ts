export enum AccountEvent {
  CREATED = "account_created",
  EMAIL_CHANGED = "account_email_changed",
  OTP_CHANGED = "account_otp_changed",
  PASSWORD_CHANGED = "account_password_changed",
  PERMISSION_CHANGED = "account_permission_changed",
}
