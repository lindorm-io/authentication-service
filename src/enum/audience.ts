export enum Audience {
  ACCESS = "access",
  AUTHORIZATION = "authorization",
  IDENTITY = "identity",
  MULTI_FACTOR = "multi-factor",
  REFRESH = "refresh",
}
