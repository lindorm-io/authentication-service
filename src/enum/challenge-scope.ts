export enum ChallengeScope {
  CHANGE = "change",
  SIGN_IN = "sign_in",
  RECOVERY = "recovery",
}
