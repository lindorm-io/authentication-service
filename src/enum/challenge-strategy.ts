export enum ChallengeStrategy {
  IMPLICIT = "implicit",
  PIN = "pin",
  SECRET = "secret",
}
