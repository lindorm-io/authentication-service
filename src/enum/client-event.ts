export enum ClientEvent {
  CREATED = "client_created",
  APPROVED_CHANGED = "client_approved_changed",
  DESCRIPTION_CHANGED = "client_description_changed",
  EMAIL_AUTHORIZATION_URI_CHANGED = "client_email_authorization_uri_changed",
  NAME_CHANGED = "client_name_changed",
  SECRET_CHANGED = "client_secret_changed",
}
