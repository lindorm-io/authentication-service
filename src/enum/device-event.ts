export enum DeviceEvent {
  CREATED = "device_created",
  PIN_CHANGED = "device_pin_changed",
  SECRET_CHANGED = "device_secret_changed",
  NAME_CHANGED = "device_name_changed",
}
