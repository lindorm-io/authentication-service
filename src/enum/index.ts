export * from "./account-event";
export * from "./audience";
export * from "./device-event";
export * from "./grant-type";
export * from "./mongo-collection";
export * from "./multi-factor-challenge-type";
export * from "./response-type";
export * from "./session-event";
