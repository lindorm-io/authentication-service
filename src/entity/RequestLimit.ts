import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { GrantType } from "../enum";

export interface IRequestLimit extends IEntity {
  backOffUntil: Date;
  failedTries: number;
  grantType: GrantType;
  subject: string;
}

export interface IRequestLimitOptions extends IEntityBaseOptions {
  backOffUntil?: Date;
  failedTries?: number;
  grantType: GrantType;
  subject: string;
}

export class RequestLimit extends EntityBase implements IRequestLimit {
  private _backOffUntil: Date;
  private _failedTries: number;
  private _grantType: GrantType;
  private _subject: string;

  constructor(options: IRequestLimitOptions) {
    super(options);
    this._backOffUntil = options.backOffUntil || null;
    this._failedTries = options.failedTries || 1;
    this._grantType = options.grantType;
    this._subject = options.subject;
  }

  public get backOffUntil(): Date {
    return this._backOffUntil;
  }
  public set backOffUntil(backOffUntil: Date) {
    this._backOffUntil = backOffUntil;
    this._updated = new Date();
  }

  public get failedTries(): number {
    return this._failedTries;
  }
  public set failedTries(failedTries: number) {
    this._failedTries = failedTries;
    this._updated = new Date();
  }

  public get grantType(): GrantType {
    return this._grantType;
  }

  public get subject(): string {
    return this._subject;
  }

  public create(): void {
    /* intentionally left empty */
  }
}
