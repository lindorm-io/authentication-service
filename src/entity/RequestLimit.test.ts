import MockDate from "mockdate";
import { RequestLimit } from "./RequestLimit";
import { GrantType } from "../enum";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");
const date = new Date("2020-01-01 08:00:00.000");

describe("Session.ts", () => {
  let requestLimit: RequestLimit;

  beforeEach(() => {
    requestLimit = new RequestLimit({
      backOffUntil: new Date(),
      failedTries: 5,
      grantType: GrantType.DEVICE_PIN,
      subject: "test@lindorm.io",
    });
  });

  test("should have all data", () => {
    expect(requestLimit).toMatchSnapshot();
  });

  test("should have optional data", () => {
    requestLimit = new RequestLimit({
      grantType: GrantType.DEVICE_PIN,
      subject: "test@lindorm.io",
    });

    expect(requestLimit).toMatchSnapshot();
  });

  test("should get/set backOffUntil", () => {
    expect(requestLimit.backOffUntil).toStrictEqual(date);

    requestLimit.backOffUntil = new Date("2020-01-01 10:00:00.000");

    expect(requestLimit.backOffUntil).toStrictEqual(new Date("2020-01-01 10:00:00.000"));
  });

  test("should get/set failedTries", () => {
    expect(requestLimit.failedTries).toBe(5);

    requestLimit.failedTries = 100;

    expect(requestLimit.failedTries).toBe(100);
  });

  test("should get grantType", () => {
    expect(requestLimit.grantType).toMatchSnapshot();
  });

  test("should get subject", () => {
    expect(requestLimit.grantType).toMatchSnapshot();
  });
});
