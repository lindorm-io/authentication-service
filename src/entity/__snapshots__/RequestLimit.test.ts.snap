// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Session.ts should get grantType 1`] = `"pin_code"`;

exports[`Session.ts should get subject 1`] = `"pin_code"`;

exports[`Session.ts should have all data 1`] = `
RequestLimit {
  "_backOffUntil": 2020-01-01T07:00:00.000Z,
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_failedTries": 5,
  "_grantType": "pin_code",
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_subject": "test@lindorm.io",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`Session.ts should have optional data 1`] = `
RequestLimit {
  "_backOffUntil": null,
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_failedTries": 1,
  "_grantType": "pin_code",
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_subject": "test@lindorm.io",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;
